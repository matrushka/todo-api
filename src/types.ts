import type { ValidationError } from "class-validator";
import type { FastifyRequest } from "fastify";
import { RouteGenericInterface } from "fastify/types/route";
import type { User } from "./entity/User";

export type EntityValidationError = Error & {
  validationErrors: ValidationError[];
};

export type AccessTokenClaims = {
  user_id: string;
  iat: number;
  exp: number;
};

export type APIRequest<T extends RouteGenericInterface = {}> = FastifyRequest<
  T & {
    Querystring: T["Querystring"] & { accessToken?: string };
  }
> & { authenticate: () => Promise<User>; getCurrentUser: () => Promise<User | undefined> };
