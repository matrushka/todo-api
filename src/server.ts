import boot from "./api";
import { DEFAULT_PORT, DEVELOPMENT_ENV } from "./constants";

const { PORT = DEFAULT_PORT, NODE_ENV = DEVELOPMENT_ENV } = process.env;

boot(Number(PORT), NODE_ENV).catch((e) => {
  console.error(`FAILED TO BOOT: ${e}`);
  process.exit(1);
});
