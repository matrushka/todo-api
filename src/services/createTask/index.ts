import { validate } from "class-validator";
import { getConnection } from "typeorm";
import { Task } from "../../entity/Task";
import { EntityValidationError, TaskStatus } from "../../types";

export default async function createTask(props: { name: string; status?: TaskStatus }) {
  const task = new Task();
  task.name = props.name;
  if (props.status) task.status = props.status;

  const errors = await validate(task);
  if (errors.length) {
    const error = new Error("Failed Validation") as EntityValidationError;
    error.validationErrors = errors;
    throw error;
  }

  const connection = getConnection();
  const taskRepo = connection.getRepository(Task);

  return taskRepo.save(task);
}
