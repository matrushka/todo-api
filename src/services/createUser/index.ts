import { validate } from "class-validator";
import { getConnection } from "typeorm";
import { User } from "../../entity/User";
import { EntityValidationError } from "../../types";

export default async function createUser(props: { email: string; displayName?: string }) {
  const user = new User();
  user.email = props.email;
  user.displayName = props.displayName;

  const errors = await validate(user);
  if (errors.length) {
    const error = new Error("Failed Validation") as EntityValidationError;
    error.validationErrors = errors;
    throw error;
  }

  const connection = getConnection();
  const userRepo = connection.getRepository(User);

  return userRepo.save(user);
}
