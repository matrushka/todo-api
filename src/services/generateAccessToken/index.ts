import { User } from "../../entity/User";
import * as jwt from "jsonwebtoken";
import { AccessTokenClaims } from "../../types";
import getMasterSecret from "../../utils/getMasterSecret";

export default function generateAccessToken(user: User) {
  return jwt.sign({ user_id: user.id } as AccessTokenClaims, getMasterSecret(), {
    algorithm: "HS512",
    expiresIn: "1 day",
  });
}
