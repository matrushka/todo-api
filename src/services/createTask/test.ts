import createTask from ".";
import connection from "../../../test/connection";
import { TaskStatus } from "../../entity/Task";

connection.auto();

describe("services.createTask", () => {
  it("should create a task with valid inputs", async () => {
    const task = await createTask({ name: "test" });

    expect(task.id).not.toBeUndefined();
    expect(task.status).toBe(TaskStatus.TO_DO);
    expect(task.name).toBe("test");
    expect(task.createdAt).not.toBeUndefined();
    expect(task.updatedAt).not.toBeUndefined();
  });
});
