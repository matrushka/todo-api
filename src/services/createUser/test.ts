import createUser from ".";
import connection from "../../../test/connection";

connection.auto();

const VALID_EMAIL = "test@test.com";
const INVALID_EMAIL = "test.com";
const DISPLAY_NAME = "Test User";

describe("services.createUser", () => {
  it("should create a user with valid inputs", async () => {
    const user = await createUser({ email: VALID_EMAIL, displayName: DISPLAY_NAME });
    expect(user.id).not.toBeUndefined();
    expect(user.displayName).toBe(DISPLAY_NAME);
    expect(user.email).toBe(VALID_EMAIL);
  });

  it("should not create a user with an invalid email", async () => {
    await expect(createUser({ email: INVALID_EMAIL })).rejects.toThrowError("Failed Validation");
  });

  it("should not create a user for an existing email", async () => {
    await createUser({ email: VALID_EMAIL });
    await expect(createUser({ email: VALID_EMAIL })).rejects.toThrowError("duplicate key");
  });
});
