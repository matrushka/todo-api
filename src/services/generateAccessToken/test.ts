import generateAccessToken from ".";
import { User } from "../../entity/User";

const VALID_EMAIL = "test@test.com";
const DISPLAY_NAME = "Test User";

describe("services.generateAccessToken", () => {
  it("should create an access token for a provided user", async () => {
    const user = new User();
    user.id = "TEST-USER-ID";
    user.displayName = DISPLAY_NAME;
    user.email = VALID_EMAIL;

    const accessToken = generateAccessToken(user);
    expect(accessToken).toBeTruthy();
  });
});
