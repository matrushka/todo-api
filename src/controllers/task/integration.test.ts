import server from "../../../test/server";
import createUser from "../../services/createUser";
import generateAccessToken from "../../services/generateAccessToken";
import axios, { AxiosResponse } from "axios";
import createTask from "../../services/createTask";
import { TaskStatus } from "../../types";

server.auto();

const VALID_EMAIL = "test@test.com";

describe("controllers.task", () => {
  let accessToken;
  let existingTasks = [];

  beforeEach(async () => {
    existingTasks = [];
    const user = await createUser({ email: VALID_EMAIL });
    accessToken = generateAccessToken(user);

    for (let i = 0; i < 5; i += 1) {
      let status: TaskStatus = TaskStatus.TO_DO;
      if (i % 2 == 0) status = TaskStatus.IN_PROGRESS;
      if (i % 3 == 0) status = TaskStatus.COMPLETED;
      existingTasks.push(await createTask({ name: `Task #${i}`, status }));
    }
  });

  describe("GET /", () => {
    type Resp = {
      tasks: { name: string; id: string; status: TaskStatus }[];
    };

    it("should return all tasks w/o a status filter", async () => {
      const resp = await axios.get<Resp>(server.getUrl("/tasks"), { params: { accessToken } });

      expect(resp.status).toBe(200);
      expect(resp.data.tasks).toHaveLength(5);
      expect(resp.data.tasks.map((task) => task.name)).toMatchSnapshot();
      expect(resp.data.tasks.map((task) => task.status)).toMatchSnapshot();
    });

    it("should return filtered tasks when status query param is provided", async () => {
      const todoResp = await axios.get<Resp>(server.getUrl("/tasks"), {
        params: { status: TaskStatus.TO_DO },
      });
      expect(todoResp.data.tasks).toHaveLength(1);

      const inProgressResp = await axios.get<Resp>(server.getUrl("/tasks"), {
        params: { status: TaskStatus.IN_PROGRESS },
      });
      expect(inProgressResp.data.tasks).toHaveLength(2);

      const completedResp = await axios.get<Resp>(server.getUrl("/tasks"), {
        params: { status: TaskStatus.COMPLETED },
      });
      expect(completedResp.data.tasks).toHaveLength(2);
    });
  });

  describe("POST /", () => {
    type Resp = {
      task: { name: string; id: string; status: TaskStatus };
    };

    it("should create a new task", async () => {
      const resp = await axios.post<any, AxiosResponse<Resp>>(
        server.getUrl(`/tasks`),
        {
          task: {
            name: "New Task",
            status: TaskStatus.COMPLETED,
          },
        },
        { params: { accessToken } }
      );

      expect(resp.data.task).toBeDefined();
    });

    it.todo("should fail when accessToken is missing");
    it.todo("should fail when an invalid status is passed");
    it.todo("should fail when an empty name is passed");
  });

  describe("PUT /:id", () => {
    type Resp = {
      task: { name: string; id: string; status: TaskStatus };
    };

    it("should edit the task", async () => {
      const initialTask = existingTasks[0];
      const resp = await axios.put<any, AxiosResponse<Resp>>(
        server.getUrl(`/tasks/${initialTask.id}`),
        {
          task: {
            name: "EDITED TASK",
            status: TaskStatus.TO_DO,
          },
        },
        { params: { accessToken } }
      );

      expect(resp.data.task).toBeDefined();
      expect(resp.data.task.id).toBe(initialTask.id);
      expect(resp.data.task.name).not.toBe(initialTask.name);
      expect(resp.data.task.status).not.toBe(initialTask.status);
    });

    it.todo("should fail when accessToken is missing");
  });

  describe("DELETE /:id", () => {
    type Resp = {
      ok: true;
    };

    it.todo("should fail when accessToken is missing");
    it("should delete the task", async () => {
      const initialTask = existingTasks[0];
      const resp = await axios.delete<any, AxiosResponse<Resp>>(
        server.getUrl(`/tasks/${initialTask.id}`),
        { params: { accessToken } }
      );

      expect(resp.data.ok).toBe(true);
    });
  });
});
