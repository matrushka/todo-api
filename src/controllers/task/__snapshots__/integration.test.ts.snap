// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers.task GET / should return all tasks w/o a status filter 1`] = `
Array [
  "Task #0",
  "Task #1",
  "Task #2",
  "Task #3",
  "Task #4",
]
`;

exports[`controllers.task GET / should return all tasks w/o a status filter 2`] = `
Array [
  "COMPLETED",
  "TO_DO",
  "IN_PROGRESS",
  "COMPLETED",
  "IN_PROGRESS",
]
`;
